library/src/text/shaping.rs

```diff
@@ -679,6 +731,54 @@ fn track_and_space(ctx: &mut ShapingContext) {
     }
 }
 
+pub fn is_gb_style(styles: StyleChain) -> bool {
+    let lang = TextElem::lang_in(styles);
+    let region = TextElem::region_in(styles);
+
+    // Most CJK variants, including zh-CN, ja-JP, zh-SG, zh-MY use GB-style punctuation,
+    // while zh-HK and zh-TW use alternative style. We default to use GB-style.
+    !(lang == Lang::CHINESE
+        && matches!(region.as_ref().map(Region::as_str), Some("TW" | "HK")))
+}
+
+/// Calculate stretchability and shrinkability of each glyph,
+/// and CJK punctuation adjustments according to Chinese Layout Requirements.
+fn calculate_adjustability(ctx: &mut ShapingContext) {
+    let gb_style = is_gb_style(ctx.styles);
+
+    let mut glyphs = ctx.glyphs.iter_mut().peekable();
+    while let Some(glyph) = glyphs.next() {
+        glyph.adjustability = glyph.base_adjustability(gb_style);
+
+        if glyph.is_cjk_adjustable() {
+            // Only GB style needs further adjustment.
+            if !gb_style {
+                continue;
+            }
+
+            // Now we apply consecutive punctuation adjustment, specified in Chinese Layout
+            // Requirements, section 3.1.6.1 Punctuation Adjustment Space, and Japanese Layout
+            // Requirements, section 3.1 Line Composition Rules for Punctuation Marks
+            if let Some(next) = glyphs.peek_mut() {
+                let width = glyph.x_advance;
+                let delta = width / 2.0;
+                if next.is_cjk_adjustable()
+                    && (glyph.shrinkability().1 + next.shrinkability().0) >= delta
+                {
+                    let left_delta = glyph.shrinkability().1.min(delta);
+                    let right_delta = delta - left_delta;
+
+                    // Adjust left punctuation.
+                    glyph.shrink_right(left_delta);
+
+                    // Adjust right punctuation.
+                    next.shrink_left(right_delta);
```

Comment:
```suggestion
                    glyph.shrink_right(glyph.shrinkability().1.min(delta));
                    next.shrink_left(delta - left_delta);
```
I don't feel like the variables and comments actually increase clarity here